File: UCF HSPT Documents/2007/Solutions/radio.cpp

#include <stdio.h>
#include <stdlib.h>
#include <string.h>


// There are potenially several ways to solve this problem.  This solution
// simply fills a two dimensional boolean array representing whether or
// not a commercial is on for each minute of each station's programming.
// Then, we traverse the array, checking each minute to see if two or more
// stations are playing commercials at that time.
//
// This is probably not the most efficient solution in terms of time or
// memory used, but it is a simple, straightforward way to solve the
// problem.


int main(void)
{
   FILE *fp;
   char line[256];
   int numSessions, numStations;
   bool commercials[16][384];
   int i, j;
   int commercialGap, commercialLength;
   int minute;
   int commercialCount;
   int sessionNum;
   int pureCramming;
   
   // Open the input file
   fp = fopen("radio.in", "r");

   // Get the number of sessions to check
   fgets(line, sizeof(line), fp);
   sscanf(line, "%d", &numSessions);

   // For each session...
   for (sessionNum = 1; sessionNum <= numSessions; sessionNum++)
   {
      // Clear the commercials array to all zeroes (all false values)
      memset(commercials, 0, sizeof(commercials));

      // Get the number of stations
      fgets(line, sizeof(line), fp);
      sscanf(line, "%d", &numStations);

      // For each station...
      for (i = 0; i < numStations; i++)
      {
         // Read the station's schedule and set up the commercial
         // schedule in our array
         fgets(line, sizeof(line), fp);
         commercialGap = atoi(strtok(line, " \n"));
         commercialLength = atoi(strtok(NULL, " \n"));

         // Start at minute zero (6:00 PM)
         minute = 0;

         // Update the schedule all the way through midnight
         while (minute < 360)
         {
            // Skip the commercial gap (the amount of time between
            // commercials)
            minute += commercialGap;

            // Now, iterate through the array for commercialLength
            // minutes, and change the entries to true (there is a
            // commercial on at this time).  Keep track of the overall
            // minute as we go, so we don't overrun our array.
            j = 0;
            while ((j < commercialLength) && (minute < 360))
            {
               // Change this minute's entry to true
               commercials[i][minute] = true;

               // Increment the commercial time counter as well as the
               // overall minute counter
               minute++;
               j++;
            }
         }
      }

      // Now that the array is set up, scan it minute by minute to see
      // how many minutes of pure cramming we have
      pureCramming = 0;
      for (minute = 0; minute < 360; minute++)
      {
         // Start with zero commercial on for this minute
         commercialCount = 0;

         // Check each station for commercials
         for (i = 0; i < numStations; i++)
         {
            // If there's a commercial on this station, increment the
            // commercial count
            if (commercials[i][minute])
               commercialCount++;
         }

         // If there are two or more commercials on at this time, the
         // radio is off and pure cramming ensues
         if (commercialCount >= 2)
            pureCramming++;
      }

      // Now, we can print the output
      printf("Study session #%d has %d minute(s) of pure cramming."
         "  Excellent!\n", sessionNum, pureCramming);
   }
}

File: UCF HSPT Team/src/y2007/radio.java
import java.io.*;
import java.util.*;

public class radio
{
	public static void main(String args[]) throws Exception
	{
		BufferedReader in = new BufferedReader(new FileReader("radio.in"));
		
		int numStudySessions = Integer.parseInt(in.readLine());
		for(int studySession = 1; studySession <= numStudySessions; studySession++)
		{
			int numStations = Integer.parseInt(in.readLine());
			
			// Create a matrix, with one dimension representing a station and the
			// other dimension representing that stations' programming at a given minute,
			// starting at 6:00PM.
			int stations[][] = new int[numStations][360];
			
			// Read in the values for each individual station.
			for(int station = 0; station < numStations; station++)
			{
				StringTokenizer st = new StringTokenizer(in.readLine());
				
				int musicMinutes = Integer.parseInt(st.nextToken());
				int commercialMinutes = Integer.parseInt(st.nextToken());
				
				// isProgramming = 0: means that in this minute, this station has music playing.
				// isProgramming = 1: means that in this minute, this station has commercials playing.
				int isProgramming = 0;
				
				// Minute count represents the amount of minutes we're in our current "phase", meaning,
				// how long we've been playing music or commercials since the last switch.
				int minuteCount = 0;
				for(int minute = 0; minute < 360; minute++)
				{
					stations[station][minute] = isProgramming;
					
					minuteCount++;
					
					// If we're programming and this should be our last minute of programming,
					// reset the counter and change isProgramming.
					if(isProgramming == 0 && minuteCount == musicMinutes)
					{
                                                minuteCount = 0;
						isProgramming = 1;
					}
					// If we're in commercials and this should be our last minute of commercials,
					// reset the counter and change isProgramming.
					else if(isProgramming == 1 && minuteCount == commercialMinutes)
					{
						minuteCount = 0;
						isProgramming = 0;
					}
				}
			}
			
			// We start off with zero cramming time, and loop through each minute. Since we already have
			// the full schedule planned out in the stations matrix, we simply need to loop through, minute by
			// minute, and see which minutes have two or more stations playing commercials.
			int crammingTime = 0;
			for(int minute = 0; minute < 360; minute++)
			{
				// The index value will be 1 if there's a commercial playing, or zero if there is music.
				// We simply need to add up all the values for this given minute to figure out how many
				// stations are playing music.
				int numCommercials = 0;
				for(int station = 0; station < numStations; station++)
				{
					numCommercials += stations[station][minute];
				}
				
				// If we have more than two commercials playing, this is pure cramming time!
				if(numCommercials >= 2)
				{
					crammingTime++;
				}
			}
			
			System.out.println("Study session #"+studySession+" has "+crammingTime+" minute(s) of pure cramming.  Excellent!");
		}
	}	
}
